src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-utils.c

```diff
@@ -547,6 +547,13 @@ int xclmgmt_program_shell(struct xclmgmt_dev *lro)
 
 	xocl_thread_stop(lro);
 
+	/* when cmc release is eligible, release cmc prior to stop it */
+	ret = xocl_cmc_release_access(lro);
+	if (ret) {
+		mgmt_err(lro, "release CMC access failed %d", ret);
+		goto failed;
+	}
+
 	xocl_mb_stop(lro);
```

Comment:
should we make xocl_cmc_release_access() as part of xocl_mb_stop()?

Reply:
@houlz0507 think we should somehow get ride of mb_stop for blp? which means we should remove the xocl_mb_stop of those GPIO manipulation?  The current solution is to preserve previous code unchanged in case of any regression. Yes, we can move the new api into the xocl_mb_stop.

Reply:
OK, if you already have a plan, then do whatever you feel like.

Reply:
Discussed lizhi, I'd leave the name and location as is for this PR to enable the CMC operation. We'd have a better solution to handle all kinds of offline/online or freeze/free operations by a listener/register design module. So that only interested module will register themselves into the interest list, BLP/PLP/ULP download will simply call offline/online adapter, and the adapter will loop the interest list and call every callback.

Reply:
sounds good

Reply:
> sounds good

Please wait for final testing done. I will request re-review again. Appreciate that.

Reply:
Basic test done, xclbin download successful, reg of ep_cmc_mutex_00 is also readable.

Reply:
> sounds good

I redone the code a little bit, change the name to freeze and free because we eventually will have a unique interface called freeze and free for all kinds of sub IPs, like cmc, ert, etc.